src/shared/utilities/timeoutUtils.ts

```diff
@@ -0,0 +1,50 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+/**
+ * Timeout coTimeoutbject that can handle both cancellation token- and time limit-style timeout situations.
+ * @param timeoutLength Length of timeout duration (in ms)
+ */
+export class Timeout {
+    private readonly _startTime: number
+    private readonly _endTime: number
+    private readonly _timer: Promise<boolean>
+    public constructor(timeoutLength: number) {
+        this._startTime = new Date().getTime()
+        this._endTime = this._startTime + timeoutLength
+        this._timer = new Promise<boolean>((resolve) => {
+            setTimeout(() => resolve(false), timeoutLength)
+        })
+    }
+
+    /**
+     * Returns the amount of time left from the initialization of time Timeout object and with the timeoutLength
+     * Bottoms out at 0
+     */
+    public get remainingTime(): number {
+        const curr = new Date().getTime()
+
+        return (this._endTime - curr > 0 ? this._endTime - curr : 0)
```

Comment:
This calculates twice, could it give you an unexpected different calculation?
Consider math.max(x, 0) if there is such a feature, or put `endtime - curr` into a const first

Reply:
✔️